packages/framework-provider-local/src/services/read-model-registry.ts

```diff
@@ -32,27 +37,67 @@ export class ReadModelRegistry {
     const queryPromise = new Promise((resolve, reject) =>
       cursor.exec((err, docs) => {
         if (err) reject(err)
-        else resolve(docs)
+        else {
+          docs.map((doc: ReadModelEnvelope) => {
+            doc.value.boosterMetadata!.optimisticConcurrencyValue = this.generateOptimisticConcurrencyValue(doc.value)
+          })
+          resolve(docs)
+        }
       })
     )
-
     return queryPromise as Promise<Array<ReadModelEnvelope>>
   }
 
-  public async store(readModel: ReadModelEnvelope): Promise<void> {
+  public async store(readModel: ReadModelEnvelope, expectedCurrentVersion: number): Promise<void> {
+    const readModelOptimistic: ReadModelEnvelope & { uniqueKey?: string } = readModel
+    readModelOptimistic.uniqueKey = `${readModel.typeName}_${readModel.value.id}_${readModel.value.boosterMetadata?.version}`
+    if (readModelOptimistic.value.boosterMetadata?.version === 1) {
+      return this.insert(readModel)
+    }
+    return this.update(readModelOptimistic, expectedCurrentVersion)
+  }
+
+  private insert(readModel: ReadModelEnvelope): Promise<void> {
+    return new Promise((resolve, reject) => {
+      this.readModels.insert(readModel, (err: unknown) => {
+        err ? reject(err) : resolve()
+      })
+    })
+  }
+
+  private update(readModel: ReadModelEnvelope, expectedCurrentVersion: number): Promise<void> {
     return new Promise((resolve, reject) => {
       this.readModels.update(
-        //use nedb dot notation value.id to match the record (see https://github.com/louischatriot/nedb#finding-documents)
-        { typeName: readModel.typeName, 'value.id': readModel.value.id },
+        {
+          typeName: readModel.typeName,
+          'value.id': readModel.value.id,
+          $or: [
+            { 'value.boosterMetadata.optimisticConcurrencyValue': { $exists: false } },
+            { 'value.boosterMetadata.optimisticConcurrencyValue': expectedCurrentVersion },
```

Comment:
Oh! I thought NeDB didn't support optimistic concurrency... Why do we need this condition if we are already using the unique key approach?

Reply:
I'll change it to use the version as the field is not needed. About your question, the uniqueKey is only to prevent inserting the same read model twice and the version to update only when the version number is equals to the expected one. If the version is not equals to the expected one it means that the row was updated by another instance.